pkg/controller/common/finalizers.go

```
@@ -4,67 +4,62 @@ import (
+}
 
-		reqLogger.Info(fmt.Sprintf("Removing finalizer from %s", obj.GetObjectKind().GroupVersionKind().Kind))
-		finalizers.Delete(FinalizerName)
-		objectMeta.SetFinalizers(finalizers.List())
-		err = cl.Update(ctx, obj)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				// We're reconciling a stale instance. The object no longer exists, so we're done.
-				return false, nil
-			}
-			err = errors2.Wrapf(err, "Could not remove finalizer from %s/%s", objectMeta.GetNamespace(), objectMeta.GetName())
-			eventRecorder.Event(obj, core.EventTypeWarning, eventReasonFailedFinalizerRemoval, err.Error())
-			return false, err
-		}
+func AddFinalizer(ctx context.Context, obj runtime.Object, cl client.Client) error {
+	reqLogger := LogFromContext(ctx)
+	reqLogger.Info(fmt.Sprintf("Adding finalizer to %s", obj.GetObjectKind().GroupVersionKind().Kind))
+
+	objectMeta := getObjectMeta(obj)
+	finalizers := sets.NewString(objectMeta.GetFinalizers()...)
+	finalizers.Insert(FinalizerName)
+	objectMeta.SetFinalizers(finalizers.List())
 
-		return false, nil
+	err := cl.Update(ctx, obj)
+	if errors.IsNotFound(err) {
+		// Object was deleted manually before we could add the finalizer to it. This is not an error.
+		return nil
+	} else if err != nil {
+		return pkgerrors.Wrapf(err, "Could not add finalizer to %s/%s", objectMeta.GetNamespace(), objectMeta.GetName())
+	}
+	return nil
+}
 
-	} else if !finalizers.Has(FinalizerName) {
-		reqLogger.Info(fmt.Sprintf("Adding finalizer to %s", obj.GetObjectKind().GroupVersionKind().Kind))
-		finalizers.Insert(FinalizerName)
-		objectMeta.SetFinalizers(finalizers.List())
-		err = cl.Update(ctx, obj)
-		if errors.IsNotFound(err) {
-			// Object was deleted manually before we could add the finalizer to it. This is not an error.
-			return false, nil
-		}
-		return false, err
+func getObjectMeta(obj runtime.Object) meta.Object {
+	oma, ok := obj.(meta.ObjectMetaAccessor)
+	if !ok {
+		panic("object does not implement ObjectMetaAccessor")
```

Do we really want to panic here?  I'm guessing if this were to happen, we'd end up bouncing infinitely, with no way to correct the situation.

---

I want this to fail fast and hard during development. If this function panics, the developer is passing in an object that they shouldn't. Given that this function is called whenever the developer invokes `AddFinalizer()` or `HasFinalizer()`, it should panic the very first time the developer runs the operator, so I don't see it ever failing in production.

---

Sounds good.